modules/apps/29-fee/ibc_module.go

```diff
@@ -0,0 +1,167 @@
+package fee
+
+import (
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	sdkerrors "github.com/cosmos/cosmos-sdk/types/errors"
+	capabilitytypes "github.com/cosmos/cosmos-sdk/x/capability/types"
+	"github.com/cosmos/ibc-go/modules/apps/29-fee/keeper"
+	"github.com/cosmos/ibc-go/modules/apps/29-fee/types"
+	channeltypes "github.com/cosmos/ibc-go/modules/core/04-channel/types"
+	porttypes "github.com/cosmos/ibc-go/modules/core/05-port/types"
+	ibcexported "github.com/cosmos/ibc-go/modules/core/exported"
+)
+
+// IBCModule implements the ICS26 callbacks for the fee middleware given the fee keeper and the underlying application.
+type IBCModule struct {
+	keeper keeper.Keeper
+	app    porttypes.IBCModule
+}
+
+// NewIBCModule creates a new IBCModule given the keeper and underlying application
+func NewIBCModule(k keeper.Keeper, app porttypes.IBCModule) IBCModule {
+	return IBCModule{
+		keeper: k,
+		app:    app,
+	}
+}
+
+// OnChanOpenInit implements the IBCModule interface
+func (im IBCModule) OnChanOpenInit(
+	ctx sdk.Context,
+	order channeltypes.Order,
+	connectionHops []string,
+	portID string,
+	channelID string,
+	chanCap *capabilitytypes.Capability,
+	counterparty channeltypes.Counterparty,
+	version string,
+) error {
+	feeVersion, appVersion := channeltypes.SplitChannelVersion(version)
+	if feeVersion != "" {
+		if feeVersion != types.Version {
+			return sdkerrors.Wrapf(types.ErrInvalidVersion, "expected: %s, got: %s", types.Version, feeVersion)
+		}
+
+		im.keeper.SetFeeEnabled(ctx, portID, channelID)
+	}
+
+	// call underlying app's OnChanOpenInit callback with the appVersion
+	return im.app.OnChanOpenInit(ctx, order, connectionHops, portID, channelID,
+		chanCap, counterparty, appVersion)
+}
+
+// OnChanOpenTry implements the IBCModule interface
+func (im IBCModule) OnChanOpenTry(
+	ctx sdk.Context,
+	order channeltypes.Order,
+	connectionHops []string,
+	portID,
+	channelID string,
+	chanCap *capabilitytypes.Capability,
+	counterparty channeltypes.Counterparty,
+	version,
+	counterpartyVersion string,
+) error {
+	feeVersion, appVersion := channeltypes.SplitChannelVersion(version)
+	cpFeeVersion, cpAppVersion := channeltypes.SplitChannelVersion(counterpartyVersion)
+
+	if feeVersion != "" || cpFeeVersion != "" {
+		if feeVersion != types.Version {
+			return sdkerrors.Wrapf(types.ErrInvalidVersion, "expected: %s, got: %s", types.Version, feeVersion)
+		}
+		if cpFeeVersion != feeVersion {
+			return sdkerrors.Wrapf(types.ErrInvalidVersion, "expected counterparty version: %s, got: %s", types.Version, cpFeeVersion)
+		}
+
+		im.keeper.SetFeeEnabled(ctx, portID, channelID)
+	} else if im.keeper.IsFeeEnabled(ctx, portID, channelID) {
```

Comment:
I'm pretty sure crossing hello's are not possible on INIT with auto-generated IDs. So I think I can remove this case. @colin-axner

Reply:
Turns out this is irrelevant. We don't have to check here because core IBC checks that in case of crossing hellos,

the channel params in current TRY are exact same as the params in previous INIT: https://github.com/cosmos/ibc-go/blob/main/modules/core/04-channel/keeper/handshake.go#L131

which is exactly the behaviour we want out of core IBC, so we don't have to handle that case in the application.

I fixed the incorrect godoc in this PR.